shared/src/main/scala/io/lemonlabs/uri/Path.scala

```diff
@@ -91,12 +92,32 @@ sealed trait UrlPath extends Path {
     }
   }
 
+  /** Returns this path normalized according to
+    * <a href="http://www.ietf.org/rfc/rfc3986.txt">RFC 3986</a>
+    */
+  def normalize(removeEmptyParts: Boolean = false, slashTerminated: Option[Boolean] = None): UrlPath = {
+    val normalized0 = removeDotSegments
+    val normalized1 = if (removeEmptyParts) normalized0.removeEmptyParts else normalized0
+    val normalized2 = slashTerminated.fold(normalized1)(normalized1.slashTerminated)
```

Comment:
I think we should add an argument to this method `slashEmptyPath: Boolean = true`.

I initially thought the spec was talking about all trailing slashes on paths, but I don't think that is the case and it only applies to completely empty paths. From 6.2.3 in the spec:

> defines an empty path to be equivalent to "/"

[This wikipedia article](https://en.wikipedia.org/wiki/URI_normalization) also distinguishes these two cases as 1. **Converting an empty path to a "/" path** (a normalization that preserves semantics) and 2. **Adding a trailing "/" to a non-empty path** (a normalization that usually preserves semantic)

- I think we should implement 1. (as it is mentioned in rfc3986).
- I think the `slashTerminated` here almost implements 2. but needs to ignore empty paths

Reply:
Ah, I now see you have handled the empty path case in the `Url.normalize()` method. I think it would be good to move it here and rather than doing it only if `isHttp`, we should let the user decide by passing in `slashEmptyPath` (default true). I think it should apply to more than http. From 6.2.3. in the spec:

> In general, a URI that uses the generic syntax for authority with an
>   empty path should be normalized to a path of "/"

Reply:
Ah, I didn't realize that everything in the RFC is considered generic syntax. I misinterpreted that statement to mean that the equivalence applied only to some URIs.

I feel like having two separate options for handling slash termination made the API less than obvious so I decided to create enumeration for it. Hope that's all right.